### DISCUSSION TITLE
Why does a merge commit not trigger an `on: push` workflow?

### DISCUSSION THREAD

**Discussion by uladkasach** [community/community#25812] [Actions]
<p>I’ve setup a workflow that runs <code>on: push</code> - with no conditions.</p>
<p>after merging a PR with the <code>squash and merge</code> option, the resultant commit on the master branch did not trigger the workflow!</p>
<p>I had expected that adding a commit to the master branch would be considered a “push”, regardless of “who” pushed it (in this case, the push comming from the github ui).</p>
<p>Is this the expected behavior? If not, how could i trigger that workflow on both <code>on: push</code> and when merging?</p>

**Comment by kubkon** [comment]
<p>AFAIK if you configured your workflow with something similar to</p>
<pre><code>on:
  push:
    branches:
      - master
  pull_request:
    branches:
      - master
</code></pre>
<p>it should trigger on a merge commit treating it as “on: push”. Lately, and by that I mean since approx yesterday, there is something amiss, as this also doesn’t work for me now in this <a href="https://github.com/golemfactory/g-flite/blob/master/.github/workflows/rust.yml" rel="nofollow noopener">repo</a>, and it used to work just fine in the past. Perhaps this is related to <a href="https://github.community/t/macos-vm-failing-immediately/16763">this issue</a>? Both seem to coincide in time…</p>

**Reply by jacky0wl** [reply]
I have a workflow that is triggered just the way yours is but the workflow triggers when the pull request is opened, not when merged to main, I have no Idea why

**Comment by cormacrelf** [comment]
<p><a href="https://github.community/t/github-push-actions-suddenly-not-triggering/16767">https://github.community/t5/GitHub-Actions/Github-push-actions-suddenly-not-triggering/m-p/36988#M2753</a></p>

**Comment by Yanjingzhu** [comment, marked as answer]
<p>When a merge commit push to master, it will trigger the workflow with <strong>on:push</strong>.</p>
<p>There is the incident which caused your issue. <a href="https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fwww.githubstatus.com%2Fincidents%2F42hkbtl63nmn&amp;data=02%7C01%7Cv-yjingz%40microsoft.com%7C07e0f6f094254821213108d76287613a%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637086205896666976&amp;sdata=He5DMylX%2BKeW8UqYi4cY2WE2w%2FqpJKjcB2m5LQe2pSE%3D&amp;reserved=0" rel="nofollow noopener">https://www.githubstatus.com/incidents/42hkbtl63nmn</a>. We have resolved it. Can you check whether it works fine in your side now?</p>

**Reply by totoprayogo1916** [reply]
Doesn't work if pushed by bot, like github-actions.

**Reply by artemgavrilov** [reply]
It looks like we faced same issue. We've setup [Enable auto-merge on a pull request](https://docs.github.com/en/code-security/dependabot/working-with-dependabot/automating-dependabot-with-github-actions#enable-auto-merge-on-a-pull-request) and then we discovered that such merges don't trigger CI on `main` branch at all. There are no errors or anything else, CI workflows just not started at all.

**Reply by ekampf** [reply]
@artemgavrilov where you able to solve?

**Reply by artemgavrilov** [reply]
Hi @ekampf , this problem occurs because of [this restriction](https://docs.github.com/en/actions/using-workflows/triggering-a-workflow#triggering-a-workflow-from-a-workflow). So solution is to not use `GITHUB_TOKEN` for enabling `auto-merge` option.

**Comment by uladkasach** [comment]
<p>all working as expected now. thank you!</p>

**Comment by intolerance** [comment]
<p>Having issues with actions triggering on merge currently. Are you guys sure it wasn’t just temp. fixed?</p>

**Comment by ngmiduc** [comment]
I have also issues when merging to master, that results in no trigger of the workflow. The weird thing is, that it works for staging  but not for prod.

staging: (when merging to the dev branch, this workflow is triggered)

```
name: Staging Deployment

on:
  workflow_dispatch:
  push:
    paths:
      - ...
    branches:
      - dev
```

prod: **(when merging the release branch to  master, this workflow is _not_ triggered)**

```
name: Prod Deployment

on:
  workflow_dispatch:
  push:
    paths:
      -  ...
    branches:
      - master
```

I don't get it somehow, the workflows are quite identical. We have a mono-repo setup where we use github actions to deploy some of our services, where master is our prod branch and dev is our staging branch

**Comment by Stig-Johnny** [comment]
I have the same problem when using paths and merge commit. There are files changed in the paths. But action does not trigger. If I change something and then do an append push it works.

**Comment by omri-shilton** [comment]
We are also experiencing this issue. we have dev and main branch. and when doing a merge from dev to main and then pushing, on: push doesnt not trigger the workflow.

**Comment by polo4096** [comment]
same here when working with merge commits and paths filtering, any update ?

**Comment by alefranz** [comment]
Bumped into this issue after refactoring workflows to trigger another workflow in same repo.
It seems that having `secrets: inherit` is enough to stop triggering the workflow on push.

Is there any known workaround?

This limitation combined with https://github.com/cli/cli/discussions/8145 it means when you need to break glass you are forced to bypass the merge queue, but given the push event is not triggered, you have no way to run a workflow after merge :|
